### ISSUE TITLE
UEFI Incorrect video mode with non normal oriented panel.

### ISSUE THREAD

**Issue by howl** [memtest86plus/memtest86plus#381]
The video output appears rotated with a non normal oriented panel, the device is a Lenovo idepad D330. It's a little special because it has the panel mounted in the non normal way.

I file the bug because some UEFI binaries achieve to display in the correct way, for example the comercial with free to use version counterpart.

If the correct col x row video output mode is selected is possible to have the display as it should but I don't know how memtest86+ check for the mode to use.

If it's possible, could be added a parameter like the usb ones to pass to memtest to force video modes? If not to be in production at least to help in develop a way to auto set the correct mode.

**Comment by debrouxl** [comment]
Thanks for the report, and for gathering information about the available modes :)

I see that there's some debug code inside https://github.com/memtest86plus/memtest86plus/blob/main/boot/efisetup.c . Are you able to change the `#define DEBUG 0` to `#define DEBUG 1`, recompile the memtest86+ binary, and test it ?

**Comment by howl** [comment]
@debrouxl I have done it with 7.00 src, not nightly.

Just booting memtest86+, the first line is echoed by GRUB2:
![photo_2024-02-20_23-04-48](https://github.com/memtest86plus/memtest86plus/assets/758086/5aa5681e-a25c-4ed3-ac18-b1bb0f0ca315)

Hitting to continue:
![photo_2024-02-20_23-04-51](https://github.com/memtest86plus/memtest86plus/assets/758086/117b5a27-b53d-4d1e-b2d1-010286a3ea45)

And then hitting to continue one more time starts memtest86+.

**Comment by howl** [comment]
Is very probably that the other mode shows 480 x 640, but just guessing because of the normal orientation of the panel. This kind of panel mount is annoying, for example using X11 instead of Wayland where you can show the refresh rate moving windows, etc.. the refresh is indeed vertical instead horizontal.

**Comment by martinwhitaker** [comment]
The character-based console modes are not available once you exit the UEFI boot services, so Memtest86+ uses a graphics mode for display. It picks the mode that has the minimum width (in pixels), provided it is at least 640 pixels wide and 400 pixels high. If there is more than one mode with the same width, it picks the first one in the list supplied by the BIOS. To find out what graphics modes are being advertised by the BIOS, try applying this patch
````
diff --git a/boot/efisetup.c b/boot/efisetup.c
index 7e0a510..54bb4b8 100644
--- a/boot/efisetup.c
+++ b/boot/efisetup.c
@@ -353,6 +353,15 @@ static efi_status_t set_screen_info_from_gop(screen_info_t *si, efi_handle_t *ha
             continue;
         }
 
+#if DEBUG
+    print_string("Mode ");
+    print_dec(mode_num);
+    print_string(" : ");
+    print_dec(info->h_resolution);
+    print_string(" x ");
+    print_dec(info->v_resolution);
+    print_string("\n");
+#endif
         if (info->h_resolution >= MIN_H_RESOLUTION
          && info->v_resolution >= MIN_V_RESOLUTION
          && info->h_resolution < best_info.h_resolution) {
````
We may need to insert some additional pauses if it scrolls off the screen.

**Comment by howl** [comment]
It has plenty of lines, no need to pause :)

![photo_2024-02-21_00-24-52](https://github.com/memtest86plus/memtest86plus/assets/758086/f6737cea-5df7-4b66-b222-13f57f691ef4)

Mode 1 (640x480) is what gets selected, but I guess Mode 5 (480x640) is what it should work for this computer.

And I'm curious about what could happen with Mode 0 (1200x1920).

**Comment by martinwhitaker** [comment]
It depends on whether the screen rotation is done in hardware or software. Memtest86+ doesn't currently support rotation in software, so we can't use mode 5. It would be interesting to see the difference between mode 0 and mode 4 - does it rotate, or does it just change the aspect ratio. You could experiment by changing line 348 in efisetup.c from
````
    for (uint32_t mode_num = 0; mode_num < mode->max_mode; mode_num++) {
````
to
````
    for (uint32_t mode_num = 0; mode_num < 1; mode_num++) {
````
and then to
````
    for (uint32_t mode_num = 4; mode_num < 5; mode_num++) {
````

**Comment by howl** [comment]
Mode 1: 640x480
![photo_2024-02-21_01-20-12](https://github.com/memtest86plus/memtest86plus/assets/758086/aacaac88-5326-4311-9201-827ebdf4b93a)

Mode 5: 480x640
![photo_2024-02-21_01-20-17](https://github.com/memtest86plus/memtest86plus/assets/758086/178fb729-a43c-4bc7-aeae-dd6da186e509)

Mode 4: 1920x1200
![photo_2024-02-21_01-20-20](https://github.com/memtest86plus/memtest86plus/assets/758086/f99ccd74-a908-4e52-b4aa-759f82f13fa1)

Didn't take picture of mode 0 , let me some minutes.

**Comment by howl** [comment]
Here is Mode 0 (1200x1920):
![photo_2024-02-21_01-33-52](https://github.com/memtest86plus/memtest86plus/assets/758086/cb4d1225-6ad1-43a0-956d-4daf2f8b41c5)

Si I suppose that setting mode 5 with the ability to software rotate the output is what this device need to display it right. This kind of devices are a pain in the ass, this is not the only one with non normal oriented mounted panel.

If you need more tests tell me.

**Comment by howl** [comment]
Forget to mention that the previous screen with the two red lines and hit to continue is displayed right in all the modes.

**Comment by martinwhitaker** [comment]
Please test the screen-rotation branch and add `rotate` on the boot command line.

**Comment by howl** [comment]
With rotate takes Mode 5: 480x640 and this is the result:
![photo_2024-02-22_03-04-33](https://github.com/memtest86plus/memtest86plus/assets/758086/e3887ec0-99d9-4e9b-94ce-8fa32fb4455a)

And I then forced Mode 1: 640x480 and this is the result:
![photo_2024-02-22_03-04-40](https://github.com/memtest86plus/memtest86plus/assets/758086/c14e314d-fd45-458f-9b42-f5dcf5e56238)

With Mode 5 fullscreen is used and with what I remember with Linux when dealing initially with this kind of devices is the way to go and adjusting stride and that parameters should give the desired output. Thanks for taking this into account, being somehow only a display issue.

**Comment by martinwhitaker** [comment]
If you enable EFI debug, what is the FB stride shown for mode 5?

**Comment by howl** [comment]
It shows "FB stride : 1920".

**Comment by howl** [comment]
I have done a test hardcoding 2560 (640*4 instead 480*4) as linelength and the result is this one:
![photo_2024-02-22_21-58-03](https://github.com/memtest86plus/memtest86plus/assets/758086/d96a7c59-8656-4d07-889a-d86f27ad1064)

Big gap at top, small gap at bottom and image get outside from the right.

**Comment by martinwhitaker** [comment]
That looks like the hardware is still set to 640x480. Maybe a BIOS bug?

**Comment by martinwhitaker** [comment]
For memtest86+ the centring is done in software, based on the screen resolution (the memtest86+ display is always 640x400 pixels in graphics mode, because we have just the one 8x16 bit-mapped font). So if we have been told the resolution is 480x640 but the real resolution is 640x480, the centring will be wrong. Your last picture is exactly what I would expect after the line length has been adjusted.

**Comment by howl** [comment]
Just guessing, but with the hint you have tell how memtest do it in software, could be that the glyphs size is calculated with the opposite sizes? I mean the width should be taking into account 640 and the height 480, in the output cpu name for example is cut at Silver N50, being that position about the proportion of 480 for a 640 width, is not the half wich should be about Intel(R) P. Also the proportion is stretched is not the half, is a little less.

I'm not sure if I saying a dumb thing or could be something useful.

**Comment by martinwhitaker** [comment]
If you look at your earlier picture of mode 0 without rotation, which is 1:1 pixel mapping, so undistorted, you can see that Silver N50 ends 3/4 way across the screen, which is right for 480/640. memtest86+ does no scaling itself - the 8x16 glyphs are mapped exactly to an 8x16 region in the linear frame buffer. The aspect ratio distortion seen with the other modes comes from the graphics hardware scaling the chosen mode to the native 1200x1920 resolution of the panel.

I think the only way to get a sensible display on that machine is to provide a boot command line option to manually select the graphics mode, so you can pick one where the BIOS is telling the truth.

**Comment by howl** [comment]
I have done test for the other modes with the following results:

Mode 6: 600x800 (pitch manually set to 800x4)
![photo_2024-02-23_22-40-27](https://github.com/memtest86plus/memtest86plus/assets/758086/cb5a5921-85c0-4bad-bdf8-bb65e79d998a)

Mode 7: 768x1024 (pitch manually set to 1024x4)
![photo_2024-02-23_22-40-33](https://github.com/memtest86plus/memtest86plus/assets/758086/d1b46894-5c59-4223-b405-ef8d8d73e50b)

Mode 4: 1920x1200 (pitch manually set to 1200*4 is interesting we need to take always the second res value for the pitch to get the correct linelength)
![photo_2024-02-23_22-40-37](https://github.com/memtest86plus/memtest86plus/assets/758086/f8f02e6d-9c15-4118-8f81-9272e2839745)

So taking these results actually the best is leave the code as is because despite the output is streched and rotated this kind of devices tend to be easy to be physically rotated.

**Comment by howl** [comment]
I decided to do one more test with Mode 0 : 1200x1920 with rotate:
![photo_2024-02-23_23-04-47](https://github.com/memtest86plus/memtest86plus/assets/758086/4f956265-bed6-436e-a68d-4c548ea10b3e)

This could be the best approach because the output is centred and aspect ratio seems to be correct, but, the biggest issue is that is very small and some people could have for sure issues to see it right.

Just for the record I could see that the pitch for all the resolutions with the same area is always the same, so taking into account that 1200x1920 is special as is flipped over the other modes, even there is a quirk for this devices and little more from lenovo to swap the resolution values that other devices with the same problem about the rotation doesn't need.

```
                     pitch
Mode 0 : 1200x1920   1200*4
Mode 1 :   640x480    640*4
Mode 2 :   800x600    800*4
Mode 3 :  1024x768   1024*4
Mode 4 : 1920x1200   1200*4
Mode 5 :   480x640    640*4
Mode 6 :   600x800    800*4
Mode 7 :  768x1024   1024*4
```

**Comment by martinwhitaker** [comment]
Please retest the screen-rotation branch, this time adding the boot command line options
````
screen.mode=1024x768 screen.rhs-up
````

**Comment by howl** [comment]
This is the output, forget to set debug but will try again and post the output:
![photo_2024-02-24_04-09-41](https://github.com/memtest86plus/memtest86plus/assets/758086/aea15859-c5ed-4664-b77f-994b1705a2d2)

**Comment by howl** [comment]
Seems no to take the passed resolution, so I have double checked that the boot argument is the one you asked for.
![photo_2024-02-24_04-20-59](https://github.com/memtest86plus/memtest86plus/assets/758086/16548317-3a63-451f-8da5-7dc0ac6557e5)

**Comment by martinwhitaker** [comment]
Strange, it's working for me. I've added a bit more debug output - could you update and try again with debug enabled.

**Comment by howl** [comment]
![photo_2024-02-24_16-28-03](https://github.com/memtest86plus/memtest86plus/assets/758086/0943665a-49c6-4adb-90fb-fb45d6f68b29)
![photo_2024-02-24_16-28-06](https://github.com/memtest86plus/memtest86plus/assets/758086/d9038206-644d-4157-b95d-a9b8586c32d6)

Don't know why it doesn't take the res arguments, rhs-up is take into account. Is GRUB 2.06.

**Comment by martinwhitaker** [comment]
I have a theory. Try this patch:
````
--- a/system/screen.c
+++ b/system/screen.c
@@ -122,9 +122,6 @@ static void parse_option(const char *option, int option_length, uint32_t *w, uin
 
 static void parse_cmd_line(uintptr_t cmd_line_addr, uint32_t cmd_line_size, uint32_t *w, uint32_t *h)
 {
-    if (cmd_line_parsed)
-        return;
-
     cmd_line_parsed = true;
 
     if (cmd_line_addr != 0) {
````

**Comment by martinwhitaker** [comment]
I think that display is the best we can do unless there's some magic to make the 480x640 mode work properly. Out of curiosity, if you choose that mode with debug enabled, what does the test screen (red bars, normally top and bottom) look like?

**Comment by howl** [comment]
This device is a free soul, what is normally for it seems to be boring.

Mode 5 :   480x640
![photo_2024-02-24_19-28-31](https://github.com/memtest86plus/memtest86plus/assets/758086/0be043a7-10b2-4df0-bb0a-e2e36dae3ba6)

Mode 1 :   640x480
![photo_2024-02-24_19-28-35](https://github.com/memtest86plus/memtest86plus/assets/758086/69ad34cd-991f-4811-977a-73be2b306485)

Both modes left right red bars.

**Comment by martinwhitaker** [comment]
I wonder if the 480x640 mode just tells the BIOS to rotate the text. Unfortunately my test pattern doesn't let me tell whether the display is physically operating in 480x640 or 640x480 mode, as we can't see the individual pixels. I have altered it to only draw on alternate lines which I think would show a difference.

I've updated the screen-rotation branch to fix the earlier bug. I've also added an "efidebug" boot option to save having to recompile with DEBUG=1 in efisetup.c. Could you test one more time to check it works without any patches now.

**Comment by howl** [comment]
Mode 5 : 480x640
![photo_2024-02-25_19-36-33](https://github.com/memtest86plus/memtest86plus/assets/758086/fa4419de-2e2f-45fe-bbff-134e67345083)
![photo_2024-02-25_19-36-36](https://github.com/memtest86plus/memtest86plus/assets/758086/b2c9780d-b4a9-40a4-bde0-85e8ece3de79)
![photo_2024-02-25_19-36-40](https://github.com/memtest86plus/memtest86plus/assets/758086/c2596c62-4228-4531-a2e0-faffe2c3e754)

Mode 1 : 640x480
![photo_2024-02-25_19-36-42](https://github.com/memtest86plus/memtest86plus/assets/758086/4d12d948-a6a4-4b32-ae87-a65f6de5dc1c)

With Mode 5 the calculated pitch is 480x4 but should be 640x4, if there were just a way to determine horizontal or vertical normal orientation of the panel, but I think nobody has found a way.

For Mode 5 I have done also photos per bar because is difficult to photo the entire screen not losing the detail.

Mode 1 is just interleaved.

**Comment by howl** [comment]
Just thinking loud, doing a scale of the output could be "easy" and won't increase so much the binary size? If it's possible could be interesting for EFI graphics to take the bios native mode "given is at least 640x480" and used it then scaling the output?

I think the only mode that respect the corresponding size of the pixels is 1200x1920 and 1920x1200. Ah setting bios in the command line for the screen mode takes Mode 4 : 1920x1200.

**Comment by martinwhitaker** [comment]
OK, I think the latest pictures prove that mode 5 is still 640 pixels vertically (i.e. per line on the native panel). Because the BIOS has told us there are 480x4 bytes per line, the new test pattern fills the first 480x4 bytes in the frame buffer with red pixels, skips 480x4 bytes, fills the next 480x4 bytes with red pixels, skips 480x4 bytes, and so on. And we can see the first gap starting 3/4 of the way up the screen on the first line and ending 1/2 way up the screen on the second line.

If we filled the frame buffer with exactly the same data, mode 1 and mode 5 would look identical.

I was wondering whether the BIOS was changing something when we exited the EFI boot services that caused mode 5 to stop working properly, but I think this puts paid to that idea.

Crude scaling is easy, good scaling is a lot harder, particularly if you are starting from bit-mapped fonts.

**Comment by howl** [comment]
I have tested different video modes in GRUB2 by using for each mode:
```
terminal_output console
set gfxmode=width x height
terminal_output gfxterm
```

GRUB2 manages right the stride and the orientation, will check the source code to see how they manage that, perhaps it could be useful somehow. I will be surprised they use quirks for video output for this devices but I will be more surprised if they found a way to detect the orientation.

Also I have to test making a theme with percentage placement and sizes to check if the screen size in different modes is managed right.

**Comment by howl** [comment]
The situation is still the same for the cols rows size, because GRUB2 in graphical mode can use different font sizes and so on. I think this BIOS fail to report the correct sizes also affects GRUB2 because some resolutions that I think should have enough space GRUB complain about not having enough and refuses to set two modes (or could be that I have a custom very big font because the high DPI of the screen).

Anyway now I can say with no doubt that all the modes except 0 and 5 are bugged, the native resolution of the panel is the only one that plays fine.

**Comment by martinwhitaker** [comment]
GRUB gets the stride and orientation right because it has a shadow buffer and uses the EFI GOP BLT function to copy to the real frame buffer. So it doesn't use the PixelsPerScanLine value (because the software BLT buffer is never padded), and presumably the BIOS takes care of the orientation when it does the BLT.

Memtest86+ can't use the BLT function because it exits the EFI boot services (to free up the memory used by the BIOS and allow it to be tested). The other Memtest is an EFI application, so can and presumably does use the BLT function.

**Comment by howl** [comment]
That makes a lot of sense, even for more EFI applications over this device.

If you need and want to test anything else over this device just ping me.

**Comment by martinwhitaker** [comment]
Fixed by #383 (as much as we can).
